balancer/weightedroundrobin/weightedroundrobin.go

```diff
@@ -27,3 +39,181 @@ const Name = "weighted_round_robin"
+			b.scStates[sc] = connectivity.Idle
+			sc.Connect()
+		}
+	}
+	for a, sc := range b.subConns {
+		if weight, ok := addrWeights[a]; !ok {
+			// a was removed by resolver.
+			b.cc.RemoveSubConn(sc.SubConn)
+			delete(b.subConns, a)
+			// Keep the state of this sc in b.scStates until sc's state becomes Shutdown.
+			// The entry will be deleted in HandleSubConnStateChange.
+		} else {
+			sc.Weight = weight
+		}
+	}
+}
+
+func (b *wrrBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	s := state.ConnectivityState
+	grpclog.Infof("wrr: handle SubConn state change: %p, %v", sc, s)
+	oldS, ok := b.scStates[sc]
+	if !ok {
+		grpclog.Infof("wrr: got state changes for an unknown SubConn: %p, %v", sc, s)
+		return
+	}
+	b.scStates[sc] = s
+	switch s {
+	case connectivity.Idle:
+		sc.Connect()
+	case connectivity.Shutdown:
+		// When an address was removed by resolver, b called RemoveSubConn but
+		// kept the sc's state in scStates. Remove state for this sc here.
+		delete(b.scStates, sc)
+	}
+
+	oldAggrState := b.state
+	b.state = b.csEvltr.RecordTransition(oldS, s)
+
+	// Regenerate picker when one of the following happens:
+	//  - this sc became ready from not-ready
+	//  - this sc became not-ready from ready
+	//  - the aggregated state of balancer became TransientFailure from non-TransientFailure
+	//  - the aggregated state of balancer became non-TransientFailure from TransientFailure
+	if (s == connectivity.Ready) != (oldS == connectivity.Ready) ||
+		(b.state == connectivity.TransientFailure) != (oldAggrState == connectivity.TransientFailure) {
+		b.regeneratePicker()
+	}
+
+	b.cc.UpdateBalancerState(b.state, b.picker)
+}
+
+// regeneratePicker takes a snapshot of the balancer, and generates a picker
+// from it. The picker is
+//  - errPicker with ErrTransientFailure if the balancer is in TransientFailure,
+//  - built by the pickerBuilder with all READY SubConns otherwise.
+func (b *wrrBalancer) regeneratePicker() {
+	if b.state == connectivity.TransientFailure {
+		b.picker = base.NewErrPicker(balancer.ErrTransientFailure)
+		return
+	}
+	readySCs := make(map[string]*balancerAddrInfo)
+
+	// Filter out all ready SCs from full subConn map.
+	for addr, sc := range b.subConns {
+		if st, ok := b.scStates[sc.SubConn]; ok && st == connectivity.Ready {
+			readySCs[addr] = sc
+		}
+	}
+	b.updatePicker(readySCs)
+}
+
+func (b *wrrBalancer) updatePicker(readySCs map[string]*balancerAddrInfo) {
+	grpclog.Infof("wrr: updatePicker called with readySCs: %v", readySCs)
+	if len(readySCs) == 0 {
+		b.picker = base.NewErrPicker(balancer.ErrNoSubConnAvailable)
+		return
+	}
+	picker, ok := b.picker.(*wrrPicker)
+	if !ok {
+		picker = &wrrPicker{wrr: wrr.NewEDF()}
+	}
+	b.picker = picker
+	picker.mu.Lock()
+	w := picker.wrr
+	oldItems := w.GetItems()
+	for _, sc := range readySCs {
+		delete(oldItems, sc.SubConn)
+		w.Add(sc.SubConn, int64(sc.Weight))
+	}
+	for sc := range oldItems {
+		w.Add(sc, 0)
```

Comment:
I'm not a fan of magic value like this. I would rather add a Remove method.

And also, this is no longer `Add`, it's more like `Update`.